File: Practice/maxAreaOfInsland/maxAreaOfIsland.go
package main

func max(x int, y int) int {
	if x > y {
		return x
	}

	return y
}
func isValid(grid [][]int, visited [][]bool, i int, j int) bool {
	if i < 0 || i >= len(grid) || j < 0 || j >= len(grid[0]) || visited[i][j] || grid[i][j] == 0 {
		return false
	}

	return true
}

func dfs(grid [][]int, visited [][]bool, directions [][]int, i int, j int) int {
	area := 0
	for index := 0; index < len(directions); index++ {
		ni := i + directions[index][0]
		nj := j + directions[index][1]

		if isValid(grid, visited, ni, nj) {
			visited[ni][nj] = true
			area += 1
			area += dfs(grid, visited, directions, ni, nj)
		}
	}

	return area
}

func maxAreaOfIsland(grid [][]int) int {
	if nil == grid || len(grid) == 0 {
		return 0
	}
	directions := [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	rows := len(grid)
	columns := len(grid[0])
	visited := make([][]bool, rows)
	for i := 0; i < rows; i++ {
		visited[i] = make([]bool, columns)
	}

	maxAreaOfIsland := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < columns; j++ {
			if grid[i][j] == 1 && !visited[i][j] {
				maxAreaOfIsland = max(maxAreaOfIsland, dfs(grid, visited, directions, i, j))
			}
		}
	}

	return maxAreaOfIsland
}

File: Practice/maxAreaOfInsland/max_area_of_island.py
def is_vaiid(grid, visited, i, j):
    return not (i < 0 or i >= len(grid) or j < 0 or j >= len(grid) or visited[i][j] or grid[i][j] == 0)

def dfs(grid, visited, directions, i, j):
   area = 0
   for index in range(0, len(directions)):
       ni = i + directions[i][0]
       nj = j + directions[j][1]

       if is_vaiid(grid, visited, ni, nj):
           visited[ni][nj] = True
           area += 1
           area += dfs(grid, visited, directions, ni, nj)

    return area

def max_area_of_island(grid):
    if not grid:
        return 0
    directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    rows = len(grid)
    columns = len(grid[0])
    visited = [[False] * columns for _ in range(rows)]
    max_area_of_island = 0
    for i in range(0, rows):
        for j in range(0, columns):
            if grid[i][j] == 1 and not visited[i][j]
                visited[i][j] = True
                area = 1 + dfs(grid, visited, directions, i, j)
                max_area_of_island = max(max_area_of_island, area)
    return max_area_of_island
